src/Autofac.Pooling/IPooledRegistrationPolicy.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Generic;
+using Autofac.Core;
+
+namespace Autofac.Pooling
+{
+    /// <summary>
+    /// Defines the interface for a custom pooled registration policy, that defines what happens when an instance is retrieved from the pool, and
+    /// what happens when the instance is returned to the pool.
+    /// </summary>
+    /// <typeparam name="TPooledObject">The type of object being pooled.</typeparam>
+    public interface IPooledRegistrationPolicy<TPooledObject>
+        where TPooledObject : class
+    {
+        /// <summary>
+        /// Gets a value indicating the maximum number of items that will be retained in the pool.
```

Comment:
What "maximum number of items retained" means is not obvious. It's not a _maximum capacity_ on the pool, it's the maximum number of items that can be pooled before the pool overflows. This took me a long time to figure out.

```suggestion
        /// Gets a value indicating the maximum number of items the pool will hold before it stops
        /// tracking new instances. This is not a limit on the number of objects that can be created;
        /// it's the maximum that the pool will track.
```

Reply:
Also... how would I implement a policy with a hard limit?

Reply:
The behaviour of the `ObjectPool` is interesting in that it's not so much that it stops tracking new instances over a particular threshold, it's that when you return an object to the pool, it will throw that object away (and dispose it) if the pool is full.

The pool doesn't actually have a concept by itself of "this is how many instances are currently being used by the application".

I've updated the comment with a variation of your suggestion:

```
/// Gets a value indicating the maximum number of items that will be retained in the pool. This is not a limit
/// on the number of objects that can be created, it's the threshold past which any objects returned to 
/// the pool will be discarded.
```

There's an example of a policy that blocks threads from accessing the pool past a maximum in `ConcurrencyTests.CanUsePoolConcurrentlyWithCustomPolicyToBlockOnMaxUsage`.

That same approach could easily be adjusted to (for example) throw.

I was wondering whether that blocking policy should be added to the actual library...

Reply:
Maybe. I've always thought one of the benefits of the pool is that you can put an upper bound on the usage of a given resource.

But it doesn't have to be in this PR. Maybe we wait until someone needs it and we can add it later.